File: algorithms/[380]常数时间插入、删除和获取随机元素/solution.ts
// topics = ["数组", "哈希表"]

class RandomizedSet {
  list: number[];
  map: Map<number, number>;
  constructor() {
    // 数组存储元素值
    this.list = [];
    // 哈希表存储存储值到索引的映射
    this.map = new Map();
  }

  insert(val: number): boolean {
    if (this.map.has(val)) {
      return false;
    }

    this.list.push(val);
    this.map.set(val, this.list.length - 1);
    return true;
  }

  remove(val: number): boolean {
    if (!this.map.has(val)) {
      return false;
    }

    const idx = this.map.get(val)!;
    const len = this.list.length;

    if (len > 1) {
      // 将待删除数，交换到数组的末尾
      [this.list[idx], this.list[len - 1]] = [this.list[len - 1], this.list[idx]];
      this.map.set(this.list[idx], idx);
    }

    this.list.pop();
    this.map.delete(val);
    return true;
  }

  getRandom(): number {
    return this.list[Math.floor(Math.random() * this.list.length)];
  }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

File: algorithms/[380]常数时间插入、删除和获取随机元素/solution.py
# topics = ["设计", "数组", "哈希表"]

from random import choice
from typing import Dict, List


class RandomizedSet:
    def __init__(self):
        # 数组存储元素值
        self.list: List[int] = []
        # 哈希表存储存储值到索引的映射
        self.dict: Dict[int, int] = {}

    def insert(self, val: int) -> bool:
        if val in self.dict:
            return False
        self.list.append(val)
        self.dict[val] = len(self.list) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.dict:
            return False

        idx = self.dict[val]
        n = len(self.list)

        if n > 1:
            #  将待删除数，交换到数组的末尾
            self.list[idx], self.list[-1] = self.list[-1], self.list[idx]
            self.dict[self.list[idx]] = idx

        self.list.pop()
        self.dict.pop(val)
        return True

    def getRandom(self) -> int:
        return choice(self.list)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
